Create a field when Ctrl-dropping a resource into the code editor

JoNax97 opened the pull request:
The code editor now creates a `const` field when dropping a resource into the code editor while CTRL/CMD is held, similar to what it does for nodes.

Closes https://github.com/godotengine/godot-proposals/issues/7673

Additionally, it fixes some inconsistencies that I found while testing the different cases (see the snippet below)

```gdscript
# Dragging a single resource works just as it did before.
"res://a_folder/an_icon.svg"

# Dragging a single resource while holding CTRL now adds a constant.
const ICON := preload("res://a_folder/icon.png")

#Dragging a script resource while holding CTRL creates a constant with PascalCase naming.
const AScript = preload("res://a_script.gd")

# Dragging multiple resources and/or folders in 4.1 caused an "Expected newline after comment string" syntax error.
"res://a_folder/", "res://a_scene.tscn"

# Dragging multiple resources and/or folders to an empty line now creates valid syntax.
"res://a_folder/"
"res://a_script.gd"

# Dragging multiple resources and/or folders to a non-empty line drops everything into 1 line.
var list = [ "res://a_folder/icon.png", "res://a_scene.tscn" ]

# Dragging multiple resources and/or folders to a non-empty line while holding CTRL skips the variables and drops everything into 1 line.
var list_2 = [ preload("res://a_folder/icon.png"), preload("res://a_scene.tscn") ]

# Dragging folders while holding CTRL in 4.1 would try to preload them, causing a syntax error.
preload("res://a_folder/")

# Dragging multiple resources and/or folders while holding CTRL now only preloads valid resources.
"res://a_folder/"
const AScript = preload("res://a_script.gd")

# Dragging an internal resource in 4.1 was missing quotation marks, resulting in a syntax error.
res://a_scene.tscn::CanvasItemMaterial_rvlbh

# Dragging an internal resource now is correctly quoted.
"res://a_scene.tscn::CanvasItemMaterial_rvlbh"

# Dragging an internal resource while holding CTRL is not allowed.
# The following error toast is displayed instead:
"Preloading internal resources is not supported"
```

JoNax97 committed:
create a constant when dropping a resource into the code editor while CRTL is held, and fix some inconsistencies

mieldepoche commented:
nice work!
Some observations:

- Naming doesn't work as expected when filenames have a dot in them: ![image](https://github.com/godotengine/godot/assets/71937306/0b6fbf4c-8c6e-4d03-9cfb-f320b03c8d68)

- > Dragging multiple resources and/or folders now creates valid syntax

I think this should only happen when dropping on an empty line though, since that particular case (with the commas) is wanted so you can easily put several resource path in an array. And if you ctrl-drop them, they all get preloaded too.

|previous behavior|your pr|
|-|-|
|<video src="https://github.com/godotengine/godot/assets/71937306/812b343f-6fbb-4413-a917-bff4f837ff68">| <video src="https://github.com/godotengine/godot/assets/71937306/e0c61565-8d6a-49e0-9365-c2bf3ba76d78">|

PinkWolfCoder commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1716,47 +1716,84 @@ static String _quote_drop_data(const String &str) {
+	}
+	variable_name = variable_name.to_snake_case().validate_identifier();
+
+	if (p_use_type) {
+		return vformat("var %s: %s = preload(%s)\n", variable_name, class_name, _quote_drop_data(path));
+	} else {
+		return vformat("var %s = preload(%s)\n", variable_name, _quote_drop_data(path));
+	}
```
``` suggestion
	if (p_use_type) {
		return vformat("var %s: %s = preload(%s)\n", variable_name, class_name, _quote_drop_data(path));
	} 
	return vformat("var %s = preload(%s)\n", variable_name, _quote_drop_data(path));
```

PinkWolfCoder commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1716,47 +1716,84 @@ static String _quote_drop_data(const String &str) {
 	int row = pos.y;
 	int col = pos.x;
+	bool preload = Input::get_singleton()->is_key_pressed(Key::CTRL);
+	bool use_type = EDITOR_GET("text_editor/completion/add_type_hints");
```
I think this one could be const

AThousandShips commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1724,39 +1755,55 @@ void ScriptTextEditor::drop_data_fw(const Point2 &p_point, const Variant &p_data
+				text_to_drop += _quote_drop_data(path);
+			}
+
+			if (i < files.size() -1) {
```
```suggestion
			if (i < files.size() - 1) {
```

JoNax97 commented:
Fixed the variable name eror and restored the previous behavior (no variable, comma separated elements) for non-empty, non-whitespace lines only. This should address the use case of filling an array as described by @mieldepoche

AThousandShips commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1716,6 +1716,37 @@ static String _quote_drop_data(const String &str) {
 	return escaped.quote(using_single_quotes ? "'" : "\"");
 }
 
+String _get_preload_resource_line(const Ref<Resource> &p_resource, bool p_use_type, bool p_create_variable) {
```
```suggestion
static String _get_preload_resource_line(const Ref<Resource> &p_resource, bool p_use_type, bool p_create_variable) {
```
Make internal functions static to prevent namespace pollution, see method just above `_quote_drop_data`

mieldepoche commented:
I tested the pr and everything works as expected now! (at least I didn't find any bug).

One thing I stumbled upon is that you can drag an internal resource before saving the scene (just after creating it), so it ends up with no valid path (either `""` or `"res://a_scene.tscn::"` depending on where it is). Maybe there should be an error message when trying to drag an unsaved resource that says "can't reference an unsaved resource, please save your scene first" or something like that? (note: I *assume* that it's what's happening here)

[simplescreenrecorder-2023-09-19_10.46.10.webm](https://github.com/godotengine/godot/assets/71937306/e91f7d8e-cde0-4300-b6c5-c2856a74f4a6)

JoNax97 commented:
For the "unsaved internal resource" error, I tried both a modal warning and a toast. 

Is there a guideline about which to use? Personally, I find the toast less intrusive.
Also, what do you think about the text?

![image](https://github.com/godotengine/godot/assets/10124534/fe5a1588-e4b1-4806-b89a-927a3810a624)
![image](https://github.com/godotengine/godot/assets/10124534/3ac64c2e-4554-43fd-bfec-2af93406f4bf)

mieldepoche commented:
> Is there a guideline about which to use

I don't know.
I too find the popup more intrusive and painful (because it requires you aiming on a small button with the mouse).
But the toast can be hidden so you can miss the message.

Calinou commented:
Should a `const` be created instead of a `var()` since `preload()` is used? See https://github.com/godotengine/godot-proposals/discussions/7841.

JoNax97 commented:
If that's desirable then sure, I can implement it. 

A couple questions though:
Does it provide any concrete advantage?
Are we sure there's no use case for overwriting the variable later? 
Should the same be done for dropped node paths?

mieldepoche commented:
>  Does it provide any concrete advantage?

it prevents you from changing it by accident (at least it should), but I'm sure you know that :p
Other than that, idk.

>  Are we sure there's no use case for overwriting the variable later?

I think that this is a niche case. Most of the time, you'll preload your resource and that's it.

>  Should the same be done for dropped node paths?

You can't.

The question of casing arises though:

- I don't agree with https://github.com/godotengine/godot-proposals/issues/7673#issuecomment-1732911942_ (which was posted in the issue that this pr adress), the constant name should *not* be in all caps, although I guess it's about personal taste but I prefer to keep the constant case (all caps) for numerical constants, not Objects. (because I find `MY_OBJECT.function()` ugly). I mean converting between constant and snake case is a shortcut away but idk. Right now all of the demo projects use variables and therefore snake_case for their preloads for example. I've seen people use `const snake_case = preload(...)` on streams, etc.
- regardless of that, preloading script should use `PascalCase` for the constant name, as per the docs:

> Also use PascalCase when loading a class into a constant or a variable:
> ```gdscript
> const Weapon = preload("res://weapon.gd")
> ```

JoNax97 commented:
I feel preloading classes has become pretty much redundant with class_name.

I'm not sure it warrants an special case.

I agree with switching to const though, it encourages good practices

mieldepoche commented:
> I feel preloading classes has become pretty much redundant with class_name.

I never use it myself, but I think ideally it would be nice to comply with the style guide. I think people use it to load a type locally, to avoid global namespace pollution (at least that's what I remember being cited).

JoNax97 commented:
I think it would be good to have some input from the editor team to see if the extra complexity is warranted.

Also, as I said in the proposal:

> Now that I think about it, why would always assume that the correct action is to preload into a constant?
> 
> Isn't loading into a variable an equally valid option?
> At any rate, it makes less assumptions about user intent.

Any opinions @Calinou @YuriSizov @KoBeWi ?

KoBeWi commented:
I think preloading a script as PascalCase const is a nice feature. I actually thought about implementing it recently. Note that this is the only way to use it as an actual type, as `var` won't work.
As for other resources, it doesn't really matter. They can be var.

JoNax97 commented:
For this particular case, should we still respect type hints setting?
`const NewScript: GDScript = preload("res://new_script.gd")`

KoBeWi commented:
No, const variables don't need type hints (they are added implicitly).

JoNax97 commented:
Done. Side question: this is the generated syntax for ctrl-dropping an internal script from a scene:

```gdscript
const SceneGdScript = preload("res://scene.tscn::GDScript_5b0km")
```

And it fails with the following message: Could not find script "res://scene.tscn::GDScript_5b0km".


Is preloading an internal resource not allowed at all? Should I detect this case and give a warning instead?

KoBeWi commented:
I think loading internal resources does work in some cases, but generally it should be avoided. So a warning (probably using the toaster) makes sense in this case. You can use the `is_built_in()` method.

JoNax97 commented:
Does this apply to any resource or only internal scripts?

KoBeWi commented:
Any resource.

JoNax97 commented:
Updated OP to reflect the new cases.

KoBeWi commented:
You need to rebase the PR. When doing force-push, always try to be up-to-date with master.

mieldepoche commented:
I tested the last changes a bit and everything seem to work as expected! and I didn't find any bug.

nitpicking:
![Screenshot from 2023-09-26 12-17-18](https://github.com/godotengine/godot/assets/71937306/36ff1d99-8908-465c-9d77-901ba6ee8dfe)
Shouldn't this be `Test_ScriptTest`? If so I guess it's a bug in godot's `to_pascal_case` (which I assume you're using). Idk if that's important.

<details>
<summary>one last thing about `const`</summary>

After testing a bit, I'm actually even more convinced it's better to `const` things by default, because it make the lines shorter/less heavy (since no explicit typing is required).
It's up to personal preference I guess, so I won't fight over it, but I had one last idea: maybe const could be behind a ctrl+shift modifier? optionally displaying an help message in the same way the 2D editor does it?
I don't realize if it's bloat at this point and the PR fully solves the issue I had already.
|dropping a texture in the 2D canvas|how it would behave in the script editor|
|-|-|
|![image](https://github.com/godotengine/godot/assets/71937306/461d8f92-a0c9-40d8-b24e-9a11f8f10d2a)|![image](https://github.com/godotengine/godot/assets/71937306/cf5772e3-fccc-4c43-b16d-8be0a15d0d29)|

It could also be an editor setting.

</details>

JoNax97 commented:
Re: const vs var: I don't think this warrants the extra complexity of a setting or talking a key combination for future use. 

If using const is preferred and has no obvious drawbacks, I will just make it so. It is the official style, after all. Those of us that prefer var can change it manually.

JoNax97 commented:
If we were to keep the "var" option behind the shift key, would someone prefer to also use "load" instead of "preload"? 

That was the only real use case where var is actually needed but I don't know if there's any demand for this.

mieldepoche commented:
> would someone prefer to also use "load" instead of "preload"

I think it's really an edge case, most of the time you want to preload the resource.
afaik `preload` will load the resource on script load, and `load` will do on resource instantiation from that  script.

so like if I get things correctly:

```gdscript
# resource.gd
extends Resource

const preloaded_member = preload("...") # will be loaded when you `const MyResource = preload("resource.gd")`
var loaded_member = load("...") # will be loaded when you `var instance := MyResource.new()`
```

but postponing the loading to instantiation time is like an edge case I think, if ever done at all willingly.

To sum it up:

- I do not think that using `load` is *not* common enough to warrant an edge case in this feature, it can be changed on the rare cases where you need it,
- likewise for `const` vs `var`. I though of the additional <kbd>Shift</kbd> as a way to cover all cases, but honestly I think simple is better (unless people start complaining and proving me wrong). Manually changing `const` to `var` when needed (+ manually adding the resource type) should be enough imo.

JoNax97 commented:
Ok, I switched to `const` and decided to keep it simple for now. If feedback shows demand for optional `var`, `load()` or whatever, it can be done in a follow-up.

The PR is tested, rebased and ready to go now.

KoBeWi commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1732,39 +1765,65 @@ void ScriptTextEditor::drop_data_fw(const Point2 &p_point, const Variant &p_data
+			return;
+		}
+		if (path.is_resource_file()) {
 			EditorNode::get_singleton()->show_warning(TTR("Only resources from filesystem can be dropped."));
```
Not really related, but I think this check does not make sense. You can remove it.

KoBeWi commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1724,6 +1724,39 @@ static String _quote_drop_data(const String &str) {
+		return vformat("preload(%s)", _quote_drop_data(path));
+	}
+
+	StringName class_name = p_resource->get_class_name();
```
```suggestion
	const StringName &class_name = p_resource->get_class_name();
```

JoNax97 replied:
I actually overwrite this value a couple lines later

JoNax97 replied:
The whole method was simplified as I don't need the class name anymore

KoBeWi commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1732,39 +1765,65 @@ void ScriptTextEditor::drop_data_fw(const Point2 &p_point, const Variant &p_data
 	int row = pos.y;
 	int col = pos.x;
 
+	const bool use_type = EDITOR_GET("text_editor/completion/add_type_hints");
```
This should only be set when `drop_modifier_pressed` is true (because otherwise it's unused). I think it was fine defined where it was before.

JoNax97 commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1724,6 +1724,39 @@ static String _quote_drop_data(const String &str) {
 
+static String _get_dropped_resource_line(const Ref<Resource> &p_resource, bool p_create_field) {
+	const String &path = p_resource->get_path();
+	const bool is_script = p_resource->is_class_ptr(Script::get_class_ptr_static());
```
I think it's better to keep consistency, I did not know of the ClassDB API

JoNax97 commented:
Applied reviews and simplified the changes significantly since there's no need to support type hints nor internal resources.

KoBeWi commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1732,39 +1753,60 @@ void ScriptTextEditor::drop_data_fw(const Point2 &p_point, const Variant &p_data
-		Ref<Resource> res = d["resource"];
-		if (!res.is_valid()) {
+		Ref<Resource> resource = d["resource"];
+		if (!resource.is_valid()) {
```
```suggestion
		if (resource.is_null()) {
```

YuriSizov commented:
Thanks!

nubels commented:
> Ok, I switched to `const` and decided to keep it simple for now. If feedback shows demand for optional `var`, `load()` or whatever, it can be done in a follow-up.
> 
> The PR is tested, rebased and ready to go now.

Awesome! Thank you very much for implementing this feature. :)
